lib/kcerts/ssh.go

```diff
@@ -0,0 +1,197 @@
+func FindSSHDir() (string, error) {
+	hDir, err := homedir.Dir()
+	if err != nil {
+		return "", fmt.Errorf("could not find the home directory: %w", err)
+	}
+	sshDir := filepath.Join(hDir, SSHDir)
+	if err := os.Mkdir(sshDir, 0700); err != nil && !os.IsExist(err) {
+		return "", fmt.Errorf("could not create directory %s: %w", sshDir, err)
+	}
+	return sshDir, nil
+}
+
+// AddSSHCAToClient adds a public key to the $HOME/.ssh/known_hosts in the ssh-cert x509.1 format.
+// For each entry, it adds an additional line and does not concatenate
+func AddSSHCAToClient(publicKey ssh.PublicKey, hosts []string, sshDir string) error {
+	caPublic := string(ssh.MarshalAuthorizedKey(publicKey))
+	knownHosts := filepath.Join(sshDir, KnownHosts)
+	knownHostsFile, err := os.OpenFile(knownHosts, os.O_RDWR|os.O_CREATE|os.O_APPEND, 0644)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return fmt.Errorf("could not create known_hosts file: %w", err)
+		}
+		return err
+	}
+	defer knownHostsFile.Close()
+	existingKnownHostsContent, err := ioutil.ReadAll(knownHostsFile)
+	if err != nil {
+		return fmt.Errorf("error reading %s: %w", knownHosts, err)
+	}
+	for _, dns := range hosts {
+		// caPublic terminates with a '\n', added by ssh.MarshalAuthorizedKey
+		publicFormat := fmt.Sprintf("%s %s %s", CAPrefix, dns, caPublic)
+		if strings.Contains(string(existingKnownHostsContent), publicFormat) {
+			continue
+		}
+		_, err = knownHostsFile.WriteString(publicFormat)
+		if err != nil {
+			return fmt.Errorf("could not add key %s to known_hosts file: %w", publicFormat, err)
+		}
+	}
+	return nil
+}
+
+type SSHAgent struct {
+	PID    int    `json:"pid"`
+	Socket string `json:"sock"`
+	// Close is edited in WriteToCache, is defaulted to an empty lambda
+	Close func() `json:"-"`
+}
+
+func (a SSHAgent) Kill() error {
+	p, err := os.FindProcess(a.PID)
+	if err != nil {
+		return err
+	}
+	return p.Kill()
+}
+
+func (a SSHAgent) Valid() bool {
+	conn, err := net.Dial("unix", a.Socket)
+	defer conn.Close()
+	return err == nil
+}
+
+// FindSSHAgent Will start the ssh agent in the interactive terminal if it isn't present already as an environment variable
+// Currently only outputs the env and does not persist it across terminals
+func FindSSHAgent(store cache.Store, logger *klog.Logger) (*SSHAgent, error) {
+	// TODO(adam): sniff out dead agents and sneak pids
+	agent := FindSSHAgentFromEnv()
+	if agent != nil && agent.Valid() {
+		return agent, nil
+	}
+	agent, err := FetchSSHAgentFromCache(store)
+	if err != nil {
+		logger.Warnf("%s", err)
+	}
+	if agent != nil && agent.Valid() {
+		return agent, nil
+	}
+	newAgent, err := CreateNewSSHAgent()
+	if err != nil {
+		return nil, err
+	}
+	logger.Infof("%s", WriteAgentToCache(store, newAgent))
+	return newAgent, nil
+}
+
+// FindSSHAgentFromEnv
+func FindSSHAgentFromEnv() *SSHAgent {
+	envSSHSock := os.Getenv("SSH_AUTH_SOCK")
+	envSSHPID := os.Getenv("SSH_AGENT_PID")
+	if envSSHSock != "" && envSSHPID != "" {
```

Comment:
nit: if you invert the if, eg `if envSSHSock == "" || envSSHPID == "" { return nil }` you reduce the newting.